zkstack_cli/crates/zkstack/src/commands/chain/gateway/migrate_from_gateway.rs

```diff
@@ -0,0 +1,311 @@
+        receipt,
+        gateway_chain_config
+            .get_contracts_config()?
+            .l1
+            .diamond_proxy_addr,
+        gateway_provider,
+    )
+    .await?;
+
+    assert!(
+        !priority_ops.is_empty(),
+        "No priority op hashes were emitted during the withdrawal calls"
+    );
+
+    let last_priority_op_hash = *priority_ops.last().unwrap();
+
+    await_for_withdrawal_to_finalize(
+        &gateway_zk_client,
+        get_ethers_provider(&l1_url)?,
+        gateway_chain_config
+            .get_contracts_config()?
+            .l1
+            .diamond_proxy_addr,
+        last_priority_op_hash,
+    )
+    .await?;
+
+    let params = gateway_zk_client
+        .get_finalize_withdrawal_params(last_priority_op_hash, 0)
+        .await?;
+
+    finish_migrate_chain_from_gateway(
+        shell,
+        args.forge_args.clone(),
+        &ecosystem_config.path_to_l1_foundry(),
+        ecosystem_config
+            .get_wallets()?
+            .deployer
+            .context("Missing deployer wallet")?,
+        ecosystem_config
+            .get_contracts_config()?
+            .ecosystem_contracts
+            .bridgehub_proxy_addr,
+        chain_config.chain_id.as_u64(),
+        gateway_chain_id,
+        params,
+        l1_url.clone(),
+    )
+    .await?;
+
+    let l1_da_validator_addr = get_l1_da_validator(&chain_config)
+        .await
+        .context("l1_da_validator_addr")?;
+
+    let spinner = Spinner::new(MSG_DA_PAIR_REGISTRATION_SPINNER);
+    set_da_validator_pair(
+        shell,
+        &ecosystem_config,
+        chain_contracts_config.l1.chain_admin_addr,
+        &chain_config.get_wallets_config()?.governor,
+        chain_contracts_config.l1.diamond_proxy_addr,
+        l1_da_validator_addr,
+        chain_contracts_config
+            .l2
+            .da_validator_addr
+            .context("da_validator_addr")?,
+        &args.forge_args,
+        l1_url.clone(),
+    )
+    .await?;
+    spinner.finish();
+    Ok(())
+}
+
+const LOOK_WAITING_TIME_MS: u64 = 200;
+
+pub(crate) async fn check_whether_gw_transaction_is_finalized(
+    gateway_provider: &Client<L2>,
+    l1_provider: Arc<Provider<Http>>,
+    gateway_diamond_proxy: Address,
+    hash: H256,
+) -> anyhow::Result<bool> {
+    let Some(receipt) = gateway_provider.get_transaction_receipt(hash).await? else {
+        return Ok(false);
+    };
+
+    if receipt.l1_batch_number.is_none() {
+        return Ok(false);
+    }
+
+    let batch_number = receipt.l1_batch_number.unwrap();
+
+    if gateway_provider
+        .get_finalize_withdrawal_params(hash, 0)
+        .await
+        .is_err()
+    {
+        return Ok(false);
+    }
+
+    // TODO(X): investigate why waiting for the tx proof is not enough.
+    // This is not expected behavior.
+    let gateway_contract = ZkChainAbi::new(gateway_diamond_proxy, l1_provider);
+    Ok(gateway_contract.get_total_batches_executed().await? >= U256::from(batch_number.as_u64()))
```

Comment:
Please add a proper todo for it. 
Our watchdog is also struggling. 
And do we need manually finalize the tx? or just block execution is enough?

Reply:
I will create a task.

We need to manually finalize it in the end